### PR TITLE
fix(taps): Honor `state_partitioning_keys` declared as class attributes

### LINKED ISSUE #3631: bug: Tap with state_partitioning_keys storage behaviour disabled no longer respected
### Singer SDK Version

0.54.1

### Is this a regression?

- [x] Yes

### Python Version

3.10

### Bug scope

Taps (catalog, state, etc.)

### Operating System

Linux

### Description

`tap-aptem` has `state_partitioning_keys` storage disabled: https://github.com/Matatika/tap-aptem/blob/358564117be4e881350f4815a1a74244a5c8609e/tap_aptem/client.py#L167

After bumping the tap `singer-sdk` version to `0.54.1` we are getting the partitions contexts stored again. This results in a massive state record, impacting the performance.

A side effect of this I also saw was my Meltano Cloud pipeline run for this tap:
1. Failed (timeout)
2. Retry Failed (timeout)
3. Retry Failed - duplicate entries error

```ValueError: State file contains duplicate entries for partition: {'AwardingBodyQualificationId': 12345, 'AwardingBodyQualification': '***'}.```

Like if the tap didn't complete the run the state was partially written back, then again on the next run, and then you have duplicates.

### Link to Slack/Linen

_No response_

### PR DESCRIPTION
**Pull request by edgarrmondragon**
Closes

- https://github.com/meltano/sdk/issues/3631

## Summary by Sourcery

Ensure stream state partitioning respects class-level configuration and improve state management typing.

Bug Fixes:
- Fix state manager initialization so streams that define state_partitioning_keys as a class attribute correctly pass those keys to the state manager.

Enhancements:
- Add precise overload typing for get_state_partition_context to clarify its behavior with None and non-None contexts.

Tests:
- Add a regression test suite verifying that class-level state_partitioning_keys values, including empty and None cases, are correctly honored when computing the state partition context.

### CODE REVIEW

File: tests/core/test_streams.py [hunk 1]
```diff
@@ -776,3 +776,59 @@ def post_process(
...
+@pytest.mark.parametrize(
+    "keys,expected_context",
+    [
+        pytest.param(
+            ["parent_id"],
+            {"parent_id": 123},
+            id="single_key",
+        ),
+        pytest.param(
+            ["parent_id", "other_key"],
```
[the remark this answers is not in the document]
Comment (edgarrmondragon) [reply tests/core/test_streams.py:790]: 0db1fb34
